### DISCUSSION TITLE
UnDiscoverables Plugin

### DISCUSSION THREAD

**Discussion by Data-Monkey** [netalertx/NetAlertX#232] [Show and tell]
Hi,
so I am trying to make a plugin for all these pesky "dumb" network devices that we would like to see in the network view.

Work in progress here: [https://github.com/Data-Monkey/Pi.Alert/tree/undiscoverables_plugin](https://github.com/Data-Monkey/Pi.Alert/tree/undiscoverables_plugin)

The goal is to read a list of devices from a file or maybe a param in the frontend and generate the devices.

**Comment by Data-Monkey** [comment]
I have some question on how the plugin works. Yes, I did read the plugins readme but I am missing something.

So far my script generates the last_results.log file and the devices in there are recognised and are added to the plugin_objects table in the DB. 

What do I need to do to make them show up in the devices? I believe it has to be in the config.json file but I can't work it out.
Do I need to map them to an existing table? I tried to map to a new table but that did not work as the table does not get created. What table do I need to map them too?

thanks

**Reply by jokob-sk** [reply]
Thanks for the hard work!

I've tried to describe the mapping of plugin data into database tables in this section:

https://github.com/jokob-sk/Pi.Alert/tree/main/front/plugins#mapping-the-plugin-results-into-a-database-table

So I'm mapping it into the dhcp_leases DB table.

Have a look at the unifi import plugin json, especially these lines:
https://github.com/jokob-sk/Pi.Alert/tree/main/front/plugins#mapping-the-plugin-results-into-a-database-table
https://github.com/jokob-sk/Pi.Alert/tree/main/front/plugins#mapping-the-plugin-results-into-a-database-table

Also it might be a better idea to map these values to the current_scan DB table instead of the dhcp_leases table, however I didn't explore this approach. 

Hope this helps!

**Reply by Data-Monkey** [reply]
ok, I did look at the mapping of the results to the tables, I just did not understand what table would be the correct one to map it too.
I can follow the DHCP_Leases plugin approach or maybe try the current scan. 

Is there a description somewhere of what the different tables are doing?

**Reply by jokob-sk** [reply]
Nothing documented about the tables but a good idea. Will do once I have spare time.

The thing is, I'm reverse engineering most if not all of the app. I'd say I have 60% of understanding of the DB, about 80% of the python script.

So I haven't investigated the current_scan functionality so far

**Reply by Data-Monkey** [reply]
I just tried the current scan table but it needs a scan id and I did not see an easy way of generating the right one. 
DHCP_Lease table worked a treat. 
I think I was confused as I looked at the DHCP Leases plugin and somehow thought the table was there just for that plugin.
Now that it works as a test at least I will clean it up and add a file or UI input to it. 

Yes, the code is not the most structured code around, I can see how it takes time to understand.

Thanks for the help.

**Reply by jokob-sk** [reply]
For the input you could use the `list` setting type: https://github.com/jokob-sk/Pi.Alert/blob/941a8ef66171d32ccfedbc9d79be03a08f18aa7a/front/plugins/website_monitor/config.json#L352

or create a different input form control for a new setting type.

The `list` setting type is then resolved like this:

![image](https://github.com/jokob-sk/Pi.Alert/assets/96159884/8540de0e-26c7-4047-8bd6-01a9a32707a3)

The actual implementation can be found here: https://github.com/jokob-sk/Pi.Alert/blob/941a8ef66171d32ccfedbc9d79be03a08f18aa7a/front/settings.php#L248

**Reply by jokob-sk** [reply]
> I just tried the current scan table but it needs a scan id and I did not see an easy way of generating the right one. 

I think it's not really used and it's fine to use the number `1`

![image](https://github.com/jokob-sk/Pi.Alert/assets/96159884/7237e5c0-258b-43b4-bf3d-c0d80b819ace)

**Reply by jokob-sk** [reply]
I've added some quick docs on teh DB structure: https://github.com/jokob-sk/Pi.Alert/blob/main/docs/DATABASE.md

**Reply by Data-Monkey** [reply]
yes, I tried that as well and I can add the data.
I discovered that the plugin mapping to tables happens after the scan is finished. So adding more records to the CurrentScan table does not help, as at the start of the next scan it gets deleted (see `def save_scanned_devices` (line 1248 in pialert.py) ) . 
So the plugin data never goes anywhere. 
We would need t do the data mapping after the scan but before the processing of the scan. 

With DHCP_Leases it works.

**Reply by Data-Monkey** [reply]
I have a working version now I believe.
Is there a special way you would like me to test it or to submit a PR?
https://github.com/Data-Monkey/Pi.Alert/tree/undiscoverables_plugin

**Comment by jokob-sk** [comment]
Closing as released
